File: src/forma/core/loader.py
"""
Document and style loaders.

Loads YAML files with optional !include resolution and validates them against
their JSON Schema (YAML format, draft-07) using the jsonschema library.

The resourceType field on each document is used to discover the matching
schema file from the registry.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

import yaml

from forma.core.include_loader import load_mapping


# ---------------------------------------------------------------------------
# Schema registry
# ---------------------------------------------------------------------------

# Maps resourceType value → schema file path.
# Extend this dict when adding new content or document types.
_FORMA_PKG = Path(__file__).parents[1]  # src/forma/
_REPO_ROOT = Path(__file__).parents[3]

_SCHEMA_REGISTRY: dict[str, Path] = {
    "FormaConfig":     _FORMA_PKG / "schema" / "forma-config.schema.yaml",
    "SlideDocument":   _FORMA_PKG / "schema" / "slide-document.schema.yaml",
    "ReportDocument":  _FORMA_PKG / "schema" / "report-document.schema.yaml",
    "ProposalContent": _FORMA_PKG / "schema" / "proposal-content.schema.yaml",
}


def register_schema(resource_type: str, schema_path: Path) -> None:
    """Register an additional resourceType → schema path mapping at runtime."""
    _SCHEMA_REGISTRY[resource_type] = schema_path


def load_document(path: Path, base_dir: Path) -> dict[str, Any]:
    """
    Load a YAML mapping file (slides.yaml / report.yaml), resolving all
    !include "@file:path" tags relative to base_dir.

    Does NOT validate — call validate_document() separately if needed.

    Args:
        path:     Path to the mapping YAML file.
        base_dir: Project root; all @file references resolve relative to this.

    Returns:
        Fully-resolved dict with all !include tags substituted.
    """
    return load_mapping(path, base_dir)


def load_content(path: Path) -> dict[str, Any]:
    """
    Load a plain content.yaml file (no !include tags).

    Returns the raw dict — useful for validation or reference lookups.
    """
    if not path.exists():
        raise FileNotFoundError(f"Content file not found: {path}")
    with open(path, encoding="utf-8") as fh:
        return yaml.safe_load(fh) or {}


def load_style(path: Path) -> dict[str, Any]:
    """
    Load a style.yaml file into a plain dict.
    """
    if not path.exists():
        return {}
    with open(path, encoding="utf-8") as fh:
        return yaml.safe_load(fh) or {}


def get_schema_for(resource_type: str) -> Path | None:
    """Return the schema file path for a given resourceType, or None."""
    return _SCHEMA_REGISTRY.get(resource_type)

File: crates/forma-core/src/loader.rs
use std::path::Path;

use forma_schema::embedded;

use crate::include_loader::load_mapping;

/// Registry mapping resourceType values to schema YAML content.
pub struct SchemaRegistry;

impl SchemaRegistry {
    /// Get the schema content for a given resourceType.
    pub fn schema_content(resource_type: &str) -> Option<&'static str> {
        for (_name, content) in embedded::all() {
            if let Some(schema_type) = Self::extract_schema_type(content) {
                if schema_type == resource_type {
                    return Some(content);
                }
            }
        }
        None
    }

    fn extract_schema_type(content: &str) -> Option<String> {
        // Collect both $id and top-level title, prefer title (matches resourceType)
        let mut id: Option<String> = None;
        let mut title: Option<String> = None;
        for line in content.lines() {
            let line = line.trim();
            // Only match top-level $id (no leading whitespace)
            if line.starts_with('$') || line.starts_with("'$id'") || line.starts_with("\"$id\"") {
                if id.is_none() {
                    let val = line.splitn(2, ':').nth(1)?.trim().trim_matches(|c| c == '\'' || c == '"').to_string();
                    id = Some(val);
                }
            }
            // Only match top-level title (no leading whitespace)
            if line.starts_with("title:") && title.is_none() {
                let val = line.splitn(2, ':').nth(1)?.trim().to_string();
                if !val.is_empty() {
                    title = Some(val);
                }
            }
        }
        // Prefer title over $id — documents use PascalCase resourceType
        title.or(id)
    }
}

/// Load a YAML mapping file (slides.yaml / report.yaml), resolving !include tags.
pub fn load_document(path: &Path, base_dir: &Path) -> Result<serde_yaml::Value, crate::include_loader::IncludeError> {
    load_mapping(path, base_dir)
}

/// Load a plain content.yaml file (no !include resolution needed).
pub fn load_content(path: &Path) -> Result<serde_yaml::Value, std::io::Error> {
    if !path.exists() {
        return Err(std::io::Error::new(std::io::ErrorKind::NotFound, format!("Content file not found: {path:?}")));
    }
    let contents = std::fs::read_to_string(path)?;
    Ok(serde_yaml::from_str(&contents).unwrap_or(serde_yaml::Value::Mapping(Default::default())))
}

/// Load a style.yaml file into a plain Value.
pub fn load_style(path: &Path) -> serde_yaml::Value {
    if !path.exists() {
        return serde_yaml::Value::Mapping(Default::default());
    }
    let contents = match std::fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => return serde_yaml::Value::Mapping(Default::default()),
    };
    serde_yaml::from_str(&contents).unwrap_or(serde_yaml::Value::Mapping(Default::default()))
}

/// Load the project config (forma.yaml).
pub fn load_config(path: &Path) -> Result<crate::config::FormaConfig, Box<dyn std::error::Error>> {
    crate::config::FormaConfig::from_yaml(path)
}
